include/seqan3/core/detail/strong_type.hpp

```diff
@@ -0,0 +1,546 @@
+        --this->underlying().get();
+        return tmp;
+    }
+    //!\}
+
+    /*!\name Conversion operators.
+     * \{
+     */
+
+    //!\brief Adds explicit conversion to it's underlying type.
+    explicit operator unnamed_type() const
+        //!\cond
+        requires ((skills & skill::convert) != skill::none)
+        //!\endcond
+    {
+        return this->underlying().get();
+    }
+    //!}
+
+protected:
+
+    //!\brief Returns a reference to the associated strong_type.
+    constexpr strong_type_ & underlying() noexcept
+    {
+        return static_cast<strong_type_ &>(*this);
+    }
+
+    //!\brief Returns a reference to the associated strong_type.
+    constexpr strong_type_ const & underlying() const noexcept
+    {
+        return static_cast<strong_type_ const &>(*this);
+    }
+};
+
+/*!\brief Base class to define expressive types.
+ * \ingroup core
+ * \tparam unnamed_type The underlying type to make expressive.
+ * \tparam name_tag     A tag, giving the `unnamed_type` a meaningful name.
+ * \tparam skills       A set of skills to be add to the expressive type.
+ *
+ * \details
+ *
+ * There are many cases in which interfaces use ambiguous parameters to
+ * that can be mixed up very easily.
+ * For example, if we consider an interface, that expects a window size and a number of maximal errors to search for
+ * possible hits in a region of interest, than both values might be given in form of an unsigned integer.
+ * The following snippet shows a perfectly well designed interface:
+ *
+ * ```cpp
+ * template <forward_range_concept fwd_rng_type>
+ * bool search(fwd_rng_type & rng, unsigned const w, unsigned const e)
+ * {
+ *  ...
+ * };
+ * ```
+ *
+ * The first parameter is the window size and the last parameter defines the error threshold.
+ * But, what happens if the user accidentally switches the `window_size` with the `error` parameter.
+ * In a bigger code base this mistake can be so subtle, that it becomes hard to figure out that the
+ * search interface was used in a wrong way. Also the compiler could not detect this.
+ * In order to make this interface safe, we would have to use a strong type.
+ * A strong type is expressive in what it actually represents as a value.
+ * In our toy example we could define two strong types as follows:
+ *
+ * ```cpp
+ * using error       = detail::strong_type<unsigned, struct search_error_tag>;
+ * using window_size = detail::strong_type<unsigned, struct window_size_tag>;
+ * ```
+ * Our interface could now be changed to:
+ *
+ * ```cpp
+ * template <forward_range_concept fwd_rng_type>
+ * bool search(fwd_rng_type & rng, window_size const w, error const e)
+ * {
+ *  ...
+ * };
+ * ```
+ *
+ * Now the user is forced to pass the parameters as their named type. If the parameter order is mixed up by accident
+ * the compiler would emit an error message, since the `error` type is not convertible to the `window_size` type and
+ * vice versa.
+ *
```

Comment:
for this example and the above example, please also show what it looks like to call the interface, e.g.
```cpp
search(my_range, 1, 3);
```
and
```cpp
search(my_range, window_size{1}, error{3});
```

Reply:
👍